### PR TITLE
[2.0] File not installed if source used previously.

### PR DESCRIPTION
**Pull request by wxtim**
Address #2438 

- [x] New test created, which fails when put on 7.8.x

Response to bug reported by @jarich.

@osanders for a  code review.
@dpmatthews for a functional review.

### CODE REVIEW

**Comment by oliver-sanders** [comment]
Forward port of https://github.com/metomi/rose/pull/2440

File: metomi/rose/config_processors/fileinstall.py [hunk 1]
```diff
@@ -253,6 +253,19 @@ def _process(self, conf_tree, nodes, loc_dao, **kwargs):
...
+                    prev_dep_locs = conn.execute(
+                        "SELECT * FROM dep_names WHERE name=?", [target.name]
+                    ).fetchall()
+                    prev_dep_locs = [i[1] for i in prev_dep_locs]
+                    prev_dep_locs = [loc_dao.select(i) for i in prev_dep_locs]
+                    if (
+                        sorted([i.name for i in prev_dep_locs]) !=
+                        sorted([i.name for i in target.dep_locs])
+                    ):
+                        target.is_out_of_date = True
```
Comment (oliver-sanders) [inline metomi/rose/config_processors/fileinstall.py:268]: I think we should wrap that with a `conn.close` for safety (unless it's done somewhere else?):

```suggestion
                    try:
                        prev_dep_locs = conn.execute(
                            "SELECT * FROM dep_names WHERE name=?", [target.name]
                        ).fetchall()
                        prev_dep_locs = [i[1] for i in prev_dep_locs]
                        prev_dep_locs = [loc_dao.select(i) for i in prev_dep_locs]
                        if (
                            sorted([i.name for i in prev_dep_locs]) !=
                            sorted([i.name for i in target.dep_locs])
                        ):
                            target.is_out_of_date = True
                    finally:
                        conn.close()
```

**Review by oliver-sanders — APPROVED** [review]
Makes sense, test looks good.

File: metomi/rose/config_processors/fileinstall.py [hunk 2]
```diff
@@ -253,6 +253,22 @@ def _process(self, conf_tree, nodes, loc_dao, **kwargs):
...
+                    conn = loc_dao.get_conn()
+                    try:
+                        prev_dep_locs = conn.execute(
+                            "SELECT * FROM dep_names WHERE name=?", [target.name]
+                        ).fetchall()
+                        prev_dep_locs = [i[1] for i in prev_dep_locs]
+                        prev_dep_locs = [loc_dao.select(i) for i in prev_dep_locs]
+                        if (
+                            sorted([i.name for i in prev_dep_locs]) !=
+                            sorted([i.name for i in target.dep_locs])
```
Comment (wxtim) [inline metomi/rose/config_processors/fileinstall.py:267]:
```suggestion
                            [i.name for i in prev_dep_locs] !=
                            [i.name for i in target.dep_locs]
```

**Review by kinow — APPROVED** [review]
Looks good to me!

File: t/rose-app-run/27-fileinstall-when-sourcelist-changed.t [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+rose app-run -C $PWD -O one -i
+rose app-run -C $PWD -O two -i
+rose app-run -C $PWD -O one -i
+# Before the bugfix rose app-run would think it had already installed source 1.
+file_cmp "${TEST_KEY_BASE}-1st-install-output" "MyTargetFile" <<__INSTALLED_FILE__
+source 1 file
+__INSTALLED_FILE__
+
+
+test_teardown
```
[the remark this answers is not in the document]
Comment (oliver-sanders) [reply t/rose-app-run/27-fileinstall-when-sourcelist-changed.t:65]: https://stackoverflow.com/a/44704969

[the remark this answers is not in the document]
Comment (wxtim) [reply t/rose-app-run/27-fileinstall-when-sourcelist-changed.t:65]:
```suggestion
__INSTALLED_FILE__


test_teardown
```

**Review by oliver-sanders — APPROVED** [review]

**Comment by oliver-sanders** [comment]
Unfortunately I'm now getting failures of `t/rose-app-run/11-file-incr-2.t` in my `beta` branch, from inspection it looks like the DB might *not* be storing the sources in order which means we may be unable to trigger reinstallation off of source order changes.

**Comment by wxtim** [comment]
More sad than confused.

**Comment by wxtim** [comment]
What test do you think we ought to apply?

**Comment by oliver-sanders** [comment]
Taking a look in, if we can make it so the sources preserve their order in the DB, problem solved, otherwise we just have to skip over it like before.

Here's the evidence as best as I can express it:

```
> ~/rose/metomi/rose/config_processors/fileinstall.py(277)_process()
-> target.is_out_of_date = True
(Pdb) conn.execute("SELECT * FROM dep_names WHERE name=?",[target.name]).fetchall()
[('hello/plants.txt', 'file:///var/tmp/tmp.aZW7XWr1Ag/repos/foo/hello/plants/fern.txt'), ('hello/plants.txt', 'file:///var/tmp/tmp.aZW7XWr1Ag/repos/foo/hello/plants/heather.txt'), ('hello/plants.txt', 'file:///var/tmp/tmp.aZW7XWr1Ag/repos/foo/hello/plants/holly.txt')]
```

```ini
[file:hello/plants.txt]
source=$REPOS_ROOT/hello/plants/heather.txt
      =$REPOS_ROOT/hello/plants/fern.txt                                       
      =$REPOS_ROOT/hello/plants/holly.txt

```

**Comment by oliver-sanders** [comment]
Got it, `SELECT` preserves insertion order, `SELECT IN` does not.

**Pull request merged**